ambry-quota/src/main/java/com/github/ambry/quota/HelixStorageUsageRefresher.java

```diff
@@ -0,0 +1,198 @@
+   * @param scheduler The {@link ScheduledExecutorService} to schedule a task to periodically refresh the usage.
+   * @param config The {@link StorageQuotaConfig}.
+   */
+  public HelixStorageUsageRefresher(HelixPropertyStore<ZNRecord> helixStore, ScheduledExecutorService scheduler,
+      StorageQuotaConfig config) {
+    this.helixStore = Objects.requireNonNull(helixStore, "HelixPropertyStore can't be null");
+    this.scheduler = scheduler;
+    this.config = config;
+    subscribeToChange();
+    initialFetchAndSchedule();
+  }
+
+  /**
+   * Fetch the storage usage and schedule a task to periodically refresh the storage usage.
+   */
+  private void initialFetchAndSchedule() {
+    Runnable updater = () -> {
+      try {
+        Map<String, Map<String, Long>> initialValue =
+            fetchContainerStorageUsageFromPath(AGGREGATED_CONTAINER_STORAGE_USAGE_PATH);
+        containerStorageUsageRef.set(initialValue);
+      } catch (Exception e) {
+        logger.error("Failed to parse the container usage from znode " + AGGREGATED_CONTAINER_STORAGE_USAGE_PATH, e);
+        // If we already have a container usage map in memory, then don't replace it with empty map.
+        containerStorageUsageRef.compareAndSet(null, Collections.EMPTY_MAP);
+      }
+    };
+    updater.run();
+
+    if (scheduler != null) {
+      int initialDelay = new Random().nextInt(config.refresherPollingIntervalMs + 1);
+      scheduler.scheduleAtFixedRate(updater, initialDelay, config.refresherPollingIntervalMs, TimeUnit.MILLISECONDS);
+      logger.info(
+          "Background storage usage updater will fetch storage usage from remote starting {} ms from now and repeat with interval={} ms",
+          initialDelay, config.refresherPollingIntervalMs);
+    }
+  }
+
+  /**
+   * Subscribe to storage usage change via helix property store. It will update the in memory cache of the storage usage
+   * and invoke the callback if present.
+   */
+  private void subscribeToChange() {
+    helixStore.subscribe(AGGREGATED_CONTAINER_STORAGE_USAGE_PATH, new HelixPropertyListener() {
+      @Override
+      public void onDataChange(String path) {
+        refreshOnUpdate(path);
+      }
+
+      @Override
+      public void onDataCreate(String path) {
+        refreshOnUpdate(path);
+      }
+
+      @Override
+      public void onDataDelete(String path) {
+        // This is a no-op when a ZNRecord for aggregated storage usage is deleted.
+        logger.warn("StorageUsage is unexpectedly deleted for at path {}", path);
+      }
+    });
+  }
+
+  @Override
+  public Map<String, Map<String, Long>> getContainerStorageUsage() {
+    return Collections.unmodifiableMap(containerStorageUsageRef.get());
+  }
+
+  @Override
+  public void registerListener(Listener cb) {
+    Objects.requireNonNull(cb, "Callback has to be non-null");
+    if (!callback.compareAndSet(null, cb)) {
+      throw new IllegalStateException("Callback already registered");
```

Comment:
If only one listener can ever be registered, can we pass it in the constructor?

Reply:
It can be passed in the constructor. Basically the refresher interface will be used in StorageQuotaService, which will be implemented later. If passing to constructor makes more sense, i will change it later when implementing quota service.